Edit post button

thecatfour opened the pull request:
### Category
- [x] **Feature**
- [ ] **Bug Fix**
- [x] **Testing**
- [ ] **Documentation**
- [ ] **Other**

### Domain
- [ ] **Frontend**
- [ ] **Backend**
- [x] **Both**

### Description
This adds an edit post button that allows users to edit their own posts. It also allows moderators/owners of hubs to edit posts in their hubs. The button is rendered using the fetchPrivileges.js file to determine if it should be visible.

This PR also has many CSS file changes, which makes it appear larger than it actually is. These changes were made to better fit with the format of other CSS files in the project. Basically, most components get their own dedicated CSS module.

### Related Issue(s)
This PR fixes #46 by implementing backend functionality for editing a post as well as frontend components to make editing a post easy.

### New Additions
**edit-post-button.tsx**
This component simply determines whether or not the edit post modal should be shown on the screen.

**edit-post-form.tsx**
This component displays the current title of the post being edited. Its input fields have the current data in them by default. It has a submit changes button and a cancel button.

**edit-post-modal.tsx**
This component holds the edit post form. 

### Modifications

- **create-input.tsx:** Added an optional parameter for starting values.
- **post-summary-detailed.tsx:** Added the edit button.
- **post-summary.tsx:** Added the edit button.
- **backend/Posts/serializers.py:** Added a PostEditSerializer.
- **backend/Posts/tests.py:** Added five unit tests for edit post functionality.
- **backend/Posts/urls.py:** Added a new url pattern.
- **backend/Posts/views.py:** Added a new view.

### New Dependencies
There are no new dependencies.

### Visuals
This is a simple gif showing how the edit post button will work:

![ezgif-5-b09ecfc18b](https://github.com/user-attachments/assets/f286633f-a273-4c30-ba78-58e8dd5d749f)

This is a simple UML diagram of how the backend will interact with the frontend to edit a post:

![image](https://github.com/user-attachments/assets/b59f38d4-0e7d-4592-9e1f-241464d08ae4)

This is a use case diagram of how the components in the frontend are used by a user. Basically, the edit post components are all included in the PostSummary and PostSummaryDetailed components and do not need to be implemented anywhere individually. This means that the only way a user can interact with the edit post features is through a post component.

![Edit Posts Use Case drawio (1)](https://github.com/user-attachments/assets/ddba78c0-a1e2-45fb-abf9-af3de73cd428)

### Architecture of Your Changes
The Posts backend was modified to allow the editing of existing posts. The serializer takes in a new title and message, and if the user has sufficient permissions, then it applies the change.

The three new components are linked together to provide an easy way to edit any post. The button includes the modal which includes the form which includes several input components. By making the button the component that controls the others, you can easily display all the necessary forms to edit a post simply by including the edit button.

### Usage and Integration
This PR adds the ability to edit posts. Although it currently uses a modal to allow users to edit posts, this could be changed to a page because of the use of components.

### Testing
Tests for the new edit post features were done in backend/Posts/tests.py. The following is a coverage report of the new features:

![image](https://github.com/user-attachments/assets/6a4681e2-f68f-4897-a8b4-2c1fc961d11a)

The frontend uses the fetchPrivileges.js helper functions to determine if the edit post button should render, and the edit post form does basic input validation to make sure the updated title and message are not empty.

### Mentions
These changes are just a user experience enhancement are likely do not have much use outside of posts themselves.

thecatfour committed:
Update serializers.py

Added a new serializer for editing posts

thecatfour committed:
Update tests.py

Added new tests for the edit post view.

thecatfour committed:
Update urls.py

Added new urls for the edit post view.

thecatfour committed:
Update views.py

Added a new view for editing a post.

thecatfour committed:
Update create-post-button.tsx

CSS changes.

thecatfour committed:
Update delete-post-button.tsx

Changed the formatting of the delete button.

thecatfour committed:
Create edit-post-button.tsx

Created a new component for editing a post.

thecatfour committed:
Update like-dislike-buttons.tsx

CSS changes.

thecatfour committed:
Create post-buttons.module.css

CSS changes.

thecatfour committed:
Moved create-input-hub

Moved create-input-hub to a different directory.

thecatfour committed:
Create create-input.module.css

CSS changes.

thecatfour committed:
Moved create-input

Moved create-input to a new directory.

thecatfour committed:
Create create-post-form.module.css

CSS changes.

thecatfour committed:
Update create-post-form.tsx

Modified how the create post form inputs are handled to allow greater modularity.

thecatfour committed:
Delete create-post-input.tsx

File was moved to different directory.

thecatfour committed:
Create edit-post-form.module.css

CSS changes.

thecatfour committed:
Create edit-post-form.tsx

Added a new form to edit an existing post. The form takes a post object and a function for taking some action if the user wants to cancel their edit.

thecatfour committed:
Delete hub-input.tsx

File was moved.

thecatfour committed:
Create delete-modal.module.css

CSS changes.

thecatfour committed:
Create delete-modal.tsx

File was moved.

thecatfour committed:
Delete delete-post-modal.tsx

File was moved.

thecatfour committed:
Create edit-post-modal.module.css

CSS changes.

thecatfour committed:
Create edit-post-modal.tsx

New modal to display the edit post form. It needs to be passed a post and a function to close the modal.

thecatfour committed:
Update detailed-post.tsx

Changed where null checking takes place for the detailed post component.

thecatfour committed:
Create post-pages.module.css

CSS changes.

thecatfour committed:
Update posts.tsx

CSS changes.

thecatfour committed:
Create post-list.module.css

CSS changes.

thecatfour committed:
Update post-list.tsx

CSS changes.

thecatfour committed:
Create post-summary-detailed.module.css

CSS changes.

thecatfour committed:
Update post-summary-detailed.tsx

Modified PostSummaryDetailed to show both the delete button and the edit post button when the user has authorization for these functions.

thecatfour committed:
Create post-summary.module.css

CSS changes.

thecatfour committed:
Update post-summary.tsx

Modifies the PostSummary component to show the delete button and edit post button when the user has authorization for these functions.

thecatfour committed:
Delete posts.module.css

CSS classes were moved to individual CSS files for each component.

thecatfour committed:
Formatting changes

Slightly changed comments and how the buttons are rendered on PostSummary and PostSummaryDetailed.

thecatfour committed:
Merge https://github.com/thecatfour/RebelHub_Fork into edit-post-button

thecatfour committed:
Update url-segments.ts

Changed the function name for create post and delete post.

thecatfour committed:
Finishing changes after the merge

Changed a few files to better fit the new url-segments.ts file.

nnguyen144 commented on frontend/src/components/posts/buttons/edit-post-button.tsx:
```diff
@@ -0,0 +1,35 @@
+import { useEffect, useState } from "react";
+import styles from "./post-buttons.module.css"
+import { createPortal } from "react-dom";
+import { checkAuthorPrivileges, checkHubPrivileges } from "@/utils/fetchPrivileges";
```
Small nitpick but this line can be removed since you aren't using the checkXPrivileges in this file

nnguyen144 commented:
I've tested the functionality of the edit post button and it works perfectly. I was able to edit the post while it was a summary and while it was a detailed post. The title and description were updated in both the frontend and the backend. And the edit post button only appears when the user is either the author of the post or a moderator/owner of the hub that the post belongs too.

I think being able to edit a post that doesn't belong to you but you are the moderator/owner of the hub that the post belongs too is a feature that we potentially might not want. Users may not want their post to be edited by any other person other than themselves. So we might want to remove that feature for hub moderators and owners. If we implement moderator/community posts that are made by moderators, then it might be fine if moderators are able to edit that post.

There are features that you may want to add later on (such as an error message if the title / description is too long, or a text that states if the post has been edited) but otherwise the edit post functionality is all there.

thecatfour committed:
Changed conditions for rendering edit post button

The edit post button now only renders for the post author. The delete button still appears for mods/owners.

thecatfour commented:
I agree that mods/owners of hubs being able to edit posts should not be an intended feature. I changed the PostSummary and PostSummaryDetailed components to only show the edit post button if the user is the author of the post. Mods and owners still have the delete button render on posts in their hubs.

I also added an error message if the title of a post is too long.

thecatfour committed:
Update create-post-validations.ts

The title for a post now has a max length of 150 characters. If there are more, then the form will display an error that the user will need to correct.

thecatfour committed:
Modified Posts backend for editing posts

Mods and owners of hubs can no longer edit posts that they do not own. They can still delete posts though.

nnguyen144 commented:
After your changes, your code looks good enough to merge after one other person has reviewed it.

thecatfour committed:
Merge branch 'UNLV-CS472-672:master' into edit-post-button

thecatfour committed:
Changed how post title and message are displayed

When editing a post, changing the title and message no longer cause the page to be reloaded. This original behavior could have caused unwanted reloads. Now, the posts use states to change the title and message that are displayed.

barrioskevin commented on backend/Posts/tests.py:
```diff
@@ -20,10 +20,123 @@ def setUp(self):
+        self.assertNotEqual(post['title'], new_data['title'], "Post title changed")
+        self.assertNotEqual(post['message'], new_data['message'], "Post message changed")
+    
+    def test_owner_edit_hub_post(self):
```
is this still a valid test case? looks like you updated some assertions but idk if it's what you intended based on the comments

thecatfour replied:
I forgot to rename it like I did with the moderator editing post test case. Owners and moderators should not be able to edit posts, but I didn't change the title of the test case.

barrioskevin commented on backend/Posts/tests.py:
```diff
@@ -20,10 +20,123 @@ def setUp(self):
+        response = self.edit_user_client.patch(reverse('post-edit', kwargs={'id': response.data['id']}), new_data, format='json')
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK, "Post could not be edited by post author")
+        self.assertEqual(response.data['title'], new_data['title'], "Post title did not change")
```
did change?

thecatfour replied:
The error message prints when the response title and the title being sent do not match, which means that the post did not change despite allowing the edit. I'm pretty sure this is correct.

barrioskevin replied:
huh? you are asserting that the response [title] is equal to new data[title] new data

barrioskevin replied:
maybe im just misunderstaing the third argument in the assertion

thecatfour replied:
The third argument is the error that is displayed if the assertion is wrong.

barrioskevin commented on backend/Posts/views.py:
```diff
@@ -68,3 +68,13 @@ def get_queryset(self):
+
+# Edit a single post by its ID
+class PostEdit(generics.UpdateAPIView):
+    queryset = Post.objects.all()
```
you can filter here so that you only look for posts that YOU made, and it will make any potential unauthorized edits easier to detect.

thecatfour replied:
Changing the queryset resulted in 404 errors for the tests rather than 403 errors, which doesn't accurately represent that the user trying to edit the post doesn't have the proper permissions. Here is an image of the failed pytest:

![image](https://github.com/user-attachments/assets/82b86d0b-e04b-4af5-90fd-5850ccf2b6dc)

barrioskevin replied:
yeah i mean it's up to you if you want to leave it with the entire posts in the query set. you're potentially leaving access to that post rather then not allowing access at all. on the client side it shouldn't make a difference since we only see edit post on posts that we make, but again up to you i just thought it would make a lot of the logic in your serializer simpler. my apologizes

barrioskevin commented:
Hello,

While I do think you edit post functionality works nice and it interacts well between frontend and backend I have a few concerns about you code in the backend.

I left a few comments on some of the test cases I was concerned about, although your tests are passing it looks like the comments could be misleading. From when you updated owner/moderator edits.

Also, if a post is updated we should maybe have a tag that let's other users know that the post has been edited right? I think this is a better design since it can be kinda confusing if a post has a lot of comments or likes and then gets edited to be something completely different.

I think in you edit post view we can make the query set be all the posts that a user owns so that way we limit the scope of posts we can edit to only those we made. 

Overall I think this is nice work and I appreciate your frontend & backend integrations

Thank you for your time,
- Kevin Barrios

thecatfour committed:
Updated tests and views for comments

Tests are now more clear with what they are doing, and they query the database in all cases to check if a post was edited.

Views was changed to only include a queryset related to the user.

thecatfour committed:
Update views.py

Restored the original queryset.

thecatfour committed:
Changed Posts backend

Added a new "last_edited" field to the post model, which stores when the post was last edited. It is initially empty, and if a post is edited, the field will be changed to whenever the post was edited.

Additionally, the tests related to editing posts were changed. A new test was made for checking if an edit post request was made without a last_edited field.

thecatfour committed:
Changed Posts frontend

Changed several post components to include the new "last_edited" field. Also updated how a component is refreshed for the user.

thecatfour committed:
Created EditedHover

Created a new component that only displays when a post has been edited. If the user hovers over the (edited) text, then the date of the edit will be displayed.

thecatfour commented:
I modified the tests in the backend to more clearly show what is being tested. I also added asserts to check the last_edited field. 

I added a date field to show that a post has been edited. When a post is edited, the field is updated from null to a date object. This can be used in the frontend to display when a post has been edited.

Here is a gif of a detailed post showing that it has been edited:

![last edited](https://github.com/user-attachments/assets/9b345bd0-b6a9-40b5-a1c6-24da652daeec)

Here is a gif of a post summary showing that it has been edited:

![summary edited](https://github.com/user-attachments/assets/d4198ec3-220c-443f-8af5-bb5b522a3769)

barrioskevin reviewed (approved):
The last edited field looks solid i appreciate your efforts.

thecatfour commented:
I think I'm gonna change how the post summary looks. We should discuss it in a meeting I think.
